File: solutions.digamma.damas.jcr/src/test/java/solutions/digamma/damas/jcr/content/JcrDocumentManagerTest.java
package solutions.digamma.damas.jcr.content;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import solutions.digamma.damas.common.NotFoundException;
import solutions.digamma.damas.auth.Token;
import solutions.digamma.damas.content.Document;
import solutions.digamma.damas.content.DocumentManager;
import solutions.digamma.damas.content.FolderManager;
import solutions.digamma.damas.jcr.Mocks;
import solutions.digamma.damas.jcr.WeldTest;

import java.io.BufferedReader;
import java.io.ByteArrayInputStream;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.nio.charset.StandardCharsets;
import java.util.stream.Collectors;

/**
 * Test scenarios for {@link JcrDocumentManager}.
 *
 * Created by Ahmad on 9/2/17.
 */
public class JcrDocumentManagerTest extends WeldTest {

    private DocumentManager manager;
    private FolderManager folderManager;
    private Token token;

    @Before
    public void setUp() throws Exception {
        this.manager = inject(JcrDocumentManager.class);
        this.folderManager = inject(FolderManager.class);
        this.token = this.login.login("admin", "admin");

    }

    @After
    public void tearDown() throws Exception {
        this.login.logout(this.token);
    }


    @Test
    public void create() throws Exception {
        final String name = "test.txt";
        String rootId = this.folderManager
                .find(this.token).getObjects().iterator().next().getId();
        Document document =
                manager.create(this.token, Mocks.document(rootId, name));
        assert name.equals(document.getName()) : "Document name mismatch";
        assert rootId.equals(document.getParentId()) : "Document ID mismatch";
    }

    @Test
    public void download() throws Exception {
        final String name = "test.txt";
        String rootId = this.folderManager
                .find(this.token).getObjects().iterator().next().getId();
        Document document =
                manager.create(this.token, Mocks.document(rootId, name));
        String id = document.getId();
        String text = "Hello";
        InputStream upload =
            new ByteArrayInputStream(text.getBytes(StandardCharsets.UTF_8));
        manager.upload(token, id, upload);
        InputStream download = manager.download(token, id).getStream();
        String readText = new BufferedReader(new InputStreamReader(download))
                .lines()
                .collect(Collectors.joining("\n"));
        assert text.equals(readText) : "Downloaded vs uploaded text mismatch.";
    }

    @Test
    public void upload() throws Exception {
        final String name = "test.txt";
        String rootId = this.folderManager
                .find(this.token).getObjects().iterator().next().getId();
        Document document =
                manager.create(this.token, Mocks.document(rootId, name));
        String id = document.getId();
        InputStream input =
            new ByteArrayInputStream("Hello".getBytes(StandardCharsets.UTF_8));
        manager.upload(token, id, input);
    }

    @Test
    public void retrieve() throws Exception {
        final String name = "test.txt";
        String rootId = this.folderManager
                .find(this.token).getObjects().iterator().next().getId();
        String id = manager
                .create(this.token, Mocks.document(rootId, name)).getId();
        Document document = manager.retrieve(token, id);
        assert id.equals(document.getId()) : "Retrieved document ID mismatch.";
        assert name.equals(document.getName()) :
                "Retrieved document name mismatch.";
        assert rootId.equals(document.getParentId()) :
                "Retrieved document parent ID mismatch.";
    }

    @Test
    public void updateName() throws Exception {
        String name = "test.txt";
        String rootId = this.folderManager
                .find(this.token).getObjects().iterator().next().getId();
        String id = manager
                .create(this.token, Mocks.document(rootId, name)).getId();
        name = "new.txt";
        manager.update(this.token, id, Mocks.document(null, name));
        Document document = manager.retrieve(this.token, id);
        assert id.equals(document.getId()) : "Updated document ID mismatch.";
        assert name.equals(document.getName()) :
                "Updated document name mismatch.";
        assert rootId.equals(document.getParentId()) :
                "Updated document parent ID mismatch.";
    }

    @Test
    public void updateParent() throws Exception {
        String name = "test.txt";
        String rootId = this.folderManager
                .find(this.token).getObjects().iterator().next().getId();
        String id = manager
                .create(this.token, Mocks.document(rootId, name)).getId();
        String folderId = folderManager
                .create(this.token, Mocks.folder(rootId, "test"))
                .getId();
        manager.update(this.token, id, Mocks.document(folderId, null));
        Document document = manager.retrieve(this.token, id);
        assert id.equals(document.getId()) : "Updated document ID mismatch.";
        assert name.equals(document.getName()) :
                "Updated document name mismatch.";
        assert folderId.equals(document.getParentId()) :
                "Updated document parent ID mismatch.";
    }

    @Test
    public void updateNameAndParent() throws Exception {
        String name = "test.txt";
        String rootId = this.folderManager
                .find(this.token).getObjects().iterator().next().getId();
        String id = manager
                .create(this.token, Mocks.document(rootId, name)).getId();
        String folderId = folderManager
                .create(this.token, Mocks.folder(rootId, "test"))
                .getId();
        name = "new.txt";
        manager.update(this.token, id, Mocks.document(folderId, name));
        Document document = manager.retrieve(this.token, id);
        assert id.equals(document.getId()) : "Updated document ID mismatch.";
        assert name.equals(document.getName()) :
                "Updated document name mismatch.";
        assert folderId.equals(document.getParentId()) :
                "Updated document parent ID mismatch.";
    }

    @Test
    public void delete() throws Exception {
        String name = "test.txt";
        String rootId = this.folderManager
                .find(this.token).getObjects().iterator().next().getId();
        String id = manager
                .create(this.token, Mocks.document(rootId, name)).getId();
        manager.delete(this.token, id);
        try {
            manager.retrieve(this.token, id);
            assert false : "NotFoundException should be thrown.";
        } catch (NotFoundException e) {}
    }

    @Test
    public void find() throws Exception {
    }
}

File: solutions.digamma.damas.jcr/src/test/kotlin/solutions/digamma/damas/jcr/content/JcrDocumentManagerTest.kt
package solutions.digamma.damas.jcr.content

import org.junit.After
import org.junit.Before
import org.junit.Test
import solutions.digamma.damas.common.NotFoundException
import solutions.digamma.damas.login.Token
import solutions.digamma.damas.content.DocumentManager
import solutions.digamma.damas.content.FolderManager
import solutions.digamma.damas.jcr.Mocks
import solutions.digamma.damas.jcr.WeldTest

import java.io.BufferedReader
import java.io.ByteArrayInputStream
import java.io.InputStreamReader
import java.nio.charset.StandardCharsets
import java.util.stream.Collectors

/**
 * Test scenarios for [JcrDocumentManager].
 *
 * Created by Ahmad on 9/2/17.
 */
class JcrDocumentManagerTest : WeldTest() {

    private var manager = WeldTest.inject(JcrDocumentManager::class.java)
    private var folderManager = WeldTest.inject(FolderManager::class.java)

    @Before
    @Throws(Exception::class)
    fun setUp() {
        this.login()
    }

    @After
    @Throws(Exception::class)
    fun tearDown() {
        this.logout()
    }


    @Test
    @Throws(Exception::class)
    fun create() {
        val name = "test.txt"
        val rootId = this.folderManager.find().objects.iterator().next().id
        val document = manager.create(Mocks.document(rootId, name))
        assert(name == document.name) { "Document name mismatch" }
        assert(rootId == document.parentId) { "Document ID mismatch" }
        manager.delete(document.id)
    }

    @Test
    @Throws(Exception::class)
    fun createWithWhiteSpace() {
        val name = "test file.txt"
        val rootId = this.folderManager.find().objects.iterator().next().id
        val document = manager.create(Mocks.document(rootId, name))
        assert(name == document.name) { "Document name mismatch" }
        assert(rootId == document.parentId) { "Document ID mismatch" }
        manager.delete(document.id)
    }

    @Test
    @Throws(Exception::class)
    fun download() {
        val name = "test.txt"
        val rootId = this.folderManager
                .find().objects.iterator().next().id
        val document = manager.create(Mocks.document(rootId, name))
        val id = document.id
        val text = "Hello"
        val upload = ByteArrayInputStream(
                text.toByteArray(StandardCharsets.UTF_8))
        manager.upload(id, upload)
        val download = manager.download(id).stream
        val readText = BufferedReader(InputStreamReader(download))
                .lines()
                .collect(Collectors.joining("\n"))
        assert(text == readText) { "Downloaded vs uploaded text mismatch." }
        manager.delete(id)
    }

    @Test
    @Throws(Exception::class)
    fun upload() {
        val name = "test.txt"
        val rootId = this.folderManager.find().objects.iterator().next().id
        val document = manager.create(Mocks.document(rootId, name))
        val id = document.id
        val input = ByteArrayInputStream(
                "Hello".toByteArray(StandardCharsets.UTF_8))
        manager.upload(id, input)
        manager.delete(id)
    }

    @Test
    @Throws(Exception::class)
    fun retrieve() {
        val name = "test.txt"
        val rootId = this.folderManager.find().objects.iterator().next().id
        val id = manager.create(Mocks.document(rootId, name)).id
        val document = manager.retrieve(id)
        assert(id == document.id) { "Retrieved document ID mismatch." }
        assert(name == document.name) { "Retrieved document name mismatch." }
        assert(rootId == document.parentId) {
            "Retrieved document parent ID mismatch."
        }
        manager.delete(id)
    }

    @Test
    @Throws(Exception::class)
    fun updateName() {
        var name = "test.txt"
        val rootId = this.folderManager.find().objects.iterator().next().id
        val id = manager.create(Mocks.document(rootId, name)).id
        name = "new.txt"
        manager.update(id, Mocks.document(null, name))
        val document = manager.retrieve(id)
        assert(id == document.id) { "Updated document ID mismatch." }
        assert(name == document.name) { "Updated document name mismatch." }
        assert(rootId == document.parentId) {
            "Updated document parent ID mismatch."
        }
        manager.delete(id)
    }

    @Test
    @Throws(Exception::class)
    fun updateNameWithWhiteSpace() {
        var name = "test.txt"
        val rootId = this.folderManager.find().objects.iterator().next().id
        val id = manager.create(Mocks.document(rootId, name)).id
        name = "new name.txt"
        manager.update(id, Mocks.document(null, name))
        val document = manager.retrieve(id)
        assert(id == document.id) { "Updated document ID mismatch." }
        assert(name == document.name) { "Updated document name mismatch." }
        assert(rootId == document.parentId) {
            "Updated document parent ID mismatch."
        }
        manager.delete(id)
    }

    @Test
    @Throws(Exception::class)
    fun updateParent() {
        val name = "test.txt"
        val rootId = this.folderManager.find().objects.iterator().next().id
        val id = manager.create(Mocks.document(rootId, name)).id
        val folderId = folderManager.create(Mocks.folder(rootId, "test")).id
        manager.update(id, Mocks.document(folderId, null))
        val document = manager.retrieve(id)
        assert(id == document.id) { "Updated document ID mismatch." }
        assert(name == document.name) { "Updated document name mismatch." }
        assert(folderId == document.parentId) {
            "Updated document parent ID mismatch."
        }
        manager.delete(id)
        folderManager.delete(folderId)
    }

    @Test
    @Throws(Exception::class)
    fun updateNameAndParent() {
        var name = "test.txt"
        val rootId = this.folderManager.find().objects.iterator().next().id
        val id = manager.create(Mocks.document(rootId, name)).id
        val folderId = folderManager.create(Mocks.folder(rootId, "test")).id
        name = "new.txt"
        manager.update(id, Mocks.document(folderId, name))
        val document = manager.retrieve(id)
        assert(id == document.id) { "Updated document ID mismatch." }
        assert(name == document.name) { "Updated document name mismatch." }
        assert(folderId == document.parentId) {
            "Updated document parent ID mismatch."
        }
        manager.delete(id)
        folderManager.delete(folderId)
    }

    @Test
    @Throws(Exception::class)
    fun updateMimeType() {
        val rootId = this.folderManager.find().objects.iterator().next().id
        val mock = Mocks.document(rootId, "test.txt")
        val mtTextPlain = "text/plain"
        Mocks.on(mock.mimeType, mtTextPlain)
        val id = manager.create(mock).id
        assert(manager.retrieve(id).mimeType == mtTextPlain) {
            "Wrong mime-type on creation."
        }
        val mtTextHtml = "text/html"
        val updater = Mocks.document()
        Mocks.on(updater.mimeType, mtTextHtml)
        manager.update(id, updater)
        assert(manager.retrieve(id).mimeType == mtTextHtml) {
            "Wrong mime-type on modification."
        }
        manager.delete(id)
    }

    @Test
    @Throws(Exception::class)
    fun delete() {
        val name = "test.txt"
        val rootId = this.folderManager.find().objects.iterator().next().id
        val id = manager.create(Mocks.document(rootId, name)).id
        manager.delete(id)
        try {
            manager.retrieve(id)
            assert(false) { "NotFoundException should be thrown." }
        } catch (e: NotFoundException) {
            assert(true) {
                "${e.javaClass} thrown instead of NotFoundException."
            }
        }
    }

    @Test
    @Throws(Exception::class)
    fun find() {
    }
}